### ISSUE TITLE
Null check failed for variable optional object attribute

### ISSUE THREAD

**Issue by SergeyBear** [hashicorp/terraform#31015]
### Terraform Version
```
Terraform v1.1.9
on linux_amd64
```

### Terraform Configuration Files
terraform.tf
```terraform
terraform {
  required_version = "~> 1.1.0"
  experiments = [module_variable_optional_attrs]
}

variable "test" {
  # type = any
  type = object({
    foo = string
    bar = optional(list(string))
  })
}

locals {
  test1 = [ for item in try(var.test.bar, []) : item ]
  test2 = [ for item in lookup(var.test, "bar", []) : item ]
}
```
terraform.tfvars
```terraform
test = { foo = "foo" }
```

### Expected Behavior
Making attribute optional should not break null checks in `can()`, `try()`, `lookup()` and should behave just like setting `type=any` for variable.
```
No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│ 
│   on terraform.tf line 3, in terraform:
│    3:   experiments = [module_variable_optional_attrs]
│ 
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│ 
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
```

### Actual Behavior
All null checks failed.
```
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│ 
│   on terraform.tf line 3, in terraform:
│    3:   experiments = [module_variable_optional_attrs]
│ 
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│ 
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵
╷
│ Error: Iteration over null value
│ 
│   on terraform.tf line 14, in locals:
│   14:   test1 = [ for item in try(var.test.bar, []) : item ]
│     ├────────────────
│     │ var.test.bar is null
│ 
│ A null value cannot be used as the collection in a 'for' expression.
╵
╷
│ Error: Iteration over null value
│ 
│   on terraform.tf line 15, in locals:
│   15:   test2 = [ for item in lookup(var.test, "bar", []) : item ]
│     ├────────────────
│     │ var.test is object with 2 attributes
│ 
│ A null value cannot be used as the collection in a 'for' expression.
```

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

### Additional Context
The `optional()` is fine but I think most natural solution would be [this](https://github.com/hashicorp/terraform/issues/19898#issuecomment-1084628191), because `optional()` does not resolve problem with lacking of `description` and `default` for object attributes.

### References
[#19898](https://github.com/hashicorp/terraform/issues/19898)

**Comment by alisdair** [comment]
Thanks for reporting this. What you're seeing here is a known limitation of the current optional object attribute experiment, which we hope to address in a future release of Terraform. When a variable value is bound to an object with optional attributes, any missing attributes are filled with `null` values. This causes `try` and `lookup` to return that `null` instead of the fallback value provided.

In the mean time, I think you can work around this behaviour by using [the `coalesce` function](https://www.terraform.io/language/functions/coalesce), which returns the first of its arguments which isn't null or an empty string:

```hcl
locals {
  test3 = [ for item in coalesce(var.test.bar, []) : item ]
}
```

**Comment by SergeyBear** [comment]
Hello! I also tried using `coalesce()`, but in some cases it make things overcomplicated, so I had to use `type=any` for objects. I hope that in the next experiment this problem will be taken into account. Thank you for a quick response!

**Comment by SergeyBear** [comment]
@alisdair I gave another chance to `coalesce()` and it actually did the job quite well. So `optional()` is still good stuff :-)
Thanks again!

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
